org.osgi.test.common/src/main/java/org/osgi/test/common/CloseableBundleContext.java

```diff
@@ -0,0 +1,239 @@
+
+public class CloseableBundleContext implements AutoCloseable, BundleContext {
+
+	private final BundleContext					bundleContext;
+	private final Set<ServiceRegistration<?>>	regs		= ConcurrentHashMap.newKeySet();
+	private final Set<Object>					listeners	= ConcurrentHashMap.newKeySet();
+	private final Set<Bundle>					bundles		= ConcurrentHashMap.newKeySet();
+
+	public CloseableBundleContext(BundleContext bundleContext) {
+		this.bundleContext = bundleContext;
+	}
+
+	@Override
+	public void close() {
+		bundles.forEach(bundle -> {
+			if ((bundle.getState() & Bundle.UNINSTALLED) != Bundle.UNINSTALLED) {
+				try {
+					bundle.uninstall();
+				} catch (BundleException e) {
+					// ignore?
+				}
+			}
+		});
+		regs.forEach(sr -> {
+			try {
+				sr.unregister();
+			} catch (IllegalStateException ise) {
+				// ignore?
+			}
+		});
+		listeners.forEach(listener -> {
+			if (listener instanceof BundleListener) {
+				bundleContext.removeBundleListener((BundleListener) listener);
+			} else if (listener instanceof ServiceListener) {
+				bundleContext.removeServiceListener((ServiceListener) listener);
+			} else if (listener instanceof FrameworkListener) {
+				bundleContext.removeFrameworkListener((FrameworkListener) listener);
+			}
+		});
+	}
+
+	@Override
+	public String getProperty(String key) {
+		return bundleContext.getProperty(key);
+	}
+
+	@Override
+	public Bundle getBundle() {
+		return bundleContext.getBundle();
+	}
+
+	@Override
+	public Bundle installBundle(String location, InputStream input) throws BundleException {
+		Bundle bundle = bundleContext.installBundle(location, input);
+		bundles.add(bundle);
+		return bundle;
+	}
+
+	@Override
+	public Bundle installBundle(String location) throws BundleException {
+		Bundle bundle = bundleContext.installBundle(location);
+		bundles.add(bundle);
+		return bundle;
+	}
+
+	@Override
+	public Bundle getBundle(long id) {
+		return bundleContext.getBundle(id);
+	}
+
+	@Override
+	public Bundle[] getBundles() {
+		return bundleContext.getBundles();
+	}
+
+	@Override
+	public void addServiceListener(ServiceListener listener, String filter) throws InvalidSyntaxException {
+		bundleContext.addServiceListener(listener, filter);
+		listeners.add(listener);
+	}
+
+	@Override
+	public void addServiceListener(ServiceListener listener) {
+		bundleContext.addServiceListener(listener);
+		listeners.add(listener);
+	}
+
+	@Override
+	public void removeServiceListener(ServiceListener listener) {
+		bundleContext.removeServiceListener(listener);
+		listeners.remove(listener);
+	}
+
+	@Override
+	public void addBundleListener(BundleListener listener) {
+		bundleContext.addBundleListener(listener);
+		listeners.add(listener);
+	}
+
+	@Override
+	public void removeBundleListener(BundleListener listener) {
+		bundleContext.removeBundleListener(listener);
+		listeners.remove(listener);
+	}
+
+	@Override
+	public void addFrameworkListener(FrameworkListener listener) {
```

Comment:
Not sure why we need to handle listeners since the framework will already handle cleaning when the bundle in stopped. Or is the point of doing this to deal with cleaning up between test method invocations?

Reply:
each test invocation.